### PR TITLE
feat: dsc delta download extension

### PR DESCRIPTION
**Pull request by hhmu**
feature implemented without deletion, which is in #166

### CODE REVIEW

File: src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java [hunk 1]
```diff
@@ -78,6 +83,9 @@
     private static final String MDC_PROP_DOWNLOAD_KEYS_TYPE = "downloadedKeysType";
     private static final String MDC_PROP_DOWNLOAD_KEYS_COUNTRY = "downloadedKeysCountry";
     private static final String DOWNLOADED_TRUSTLIST_LOG_MESSAGE = "Downloaded TrustList";
+    private static final String IF_MODIFIED_SINCE_HEADER = "If-Modified-Since";
```
Comment (f11h) [inline src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java:86]:
```suggestion
```

File: src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java [hunk 2]
```diff
@@ -42,19 +42,24 @@
...
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
 import java.util.List;
 import java.util.Locale;
 import javax.validation.Valid;
 import javax.validation.constraints.Size;
 import lombok.RequiredArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
 import org.apache.commons.collections.CollectionUtils;
+import org.springframework.http.HttpStatus;
```
Comment (f11h) [inline src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java:55]:
```suggestion
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
```

File: src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java [hunk 3]
```diff
@@ -91,13 +99,43 @@
         },
         summary = "Returns the full list of trusted certificates.",
         tags = {"Trust Lists"},
+        parameters = {
+            @Parameter(
+                in = ParameterIn.HEADER,
+                name = "If-Modified-Since",
```
Comment (f11h) [inline src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java:105]:
```suggestion
                name = HttpHeaders.IF_MODIFIED_SINCE,
```

File: src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java [hunk 4]
```diff
@@ -107,10 +145,27 @@
                 ))
         })
     public ResponseEntity<List<TrustListDto>> downloadTrustList(
+        @RequestHeader(value = IF_MODIFIED_SINCE_HEADER, required = false) String ifModifiedSince,
```
Comment (f11h) [inline src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java:148]:
```suggestion
        @RequestHeader(value = HttpHeaders.IF_MODIFIED_SINCE, required = false) ZonedDateTime ifModifiedSince,
```

File: src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java [hunk 5]
```diff
@@ -107,10 +145,27 @@
...
-
+        List<TrustListDto> trustList;
+        ZonedDateTime ifModifiedDateTime = null;
+        if (ifModifiedSince != null) {
+            try {
+                ifModifiedDateTime = ZonedDateTime.parse(ifModifiedSince, dateTimeFormatter);
+            } catch (DateTimeParseException e) {
+                return ResponseEntity.status(HttpStatus.BAD_REQUEST).build();
+            }
+        }
```
Comment (f11h) [inline src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java:161]:
```suggestion
```

File: src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java [hunk 6]
```diff
@@ -107,10 +145,27 @@
...
+        if (isPaginationRequired(page,size)) {
+            trustList = trustListMapper.trustListToTrustListDto(
+                trustListService.getTrustList(ifModifiedDateTime, page, size));
+        } else {
+            trustList = trustListMapper.trustListToTrustListDto(
+                trustListService.getTrustList(ifModifiedDateTime, null, null));
+        }
```
Comment (f11h) [inline src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java:168]:
```suggestion
        if (isPaginationRequired(page,size)) {
            trustList = trustListMapper.trustListToTrustListDto(
                trustListService.getTrustList(ifModifiedSince, page, size));
        } else {
            trustList = trustListMapper.trustListToTrustListDto(
                trustListService.getTrustList(ifModifiedSince, null, null));
        }
```

File: src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java [hunk 7]
```diff
@@ -137,18 +192,41 @@
                 name = "type",
                 description = "Certificate Type to filter for",
                 required = true,
-                schema = @Schema(implementation = CertificateTypeDto.class))
+                schema = @Schema(implementation = CertificateTypeDto.class)),
+            @Parameter(
+                in = ParameterIn.HEADER,
+                name = "If-Modified-Since",
```
Comment (f11h) [inline src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java:198]:
```suggestion
                name = HttpHeaders.IF_MODIFIED_SINCE,
```

File: src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java [hunk 8]
```diff
@@ -163,13 +241,29 @@
         })
     public ResponseEntity<List<TrustListDto>> downloadTrustListFilteredByType(
         @Valid @PathVariable("type") CertificateTypeDto type,
+        @RequestHeader(value = IF_MODIFIED_SINCE_HEADER, required = false) String ifModifiedSince,
```
Comment (f11h) [inline src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java:244]:
```suggestion
        @RequestHeader(value = HttpHeaders.IF_MODIFIED_SINCE, required = false) ZonedDateTime ifModifiedSince,
```

File: src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java [hunk 9]
```diff
@@ -163,13 +241,29 @@
...
-            trustListService.getTrustList(mappedType));
+        List<TrustListDto> trustList;
+        ZonedDateTime ifModifiedDateTime = null;
+        if (ifModifiedSince != null) {
+            try {
+                ifModifiedDateTime = ZonedDateTime.parse(ifModifiedSince, dateTimeFormatter);
+            } catch (DateTimeParseException e) {
+                return ResponseEntity.status(HttpStatus.BAD_REQUEST).build();
+            }
+        }
```
Comment (f11h) [inline src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java:259]:
```suggestion
```

File: src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java [hunk 10]
```diff
@@ -163,13 +241,29 @@
...
+        if (isPaginationRequired(page,size)) {
+            trustList = trustListMapper.trustListToTrustListDto(
+                trustListService.getTrustList(mappedType, ifModifiedDateTime, page, size));
+        } else {
+            trustList = trustListMapper.trustListToTrustListDto(
+                trustListService.getTrustList(mappedType, ifModifiedDateTime, null, null));
+        }
```
Comment (f11h) [inline src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java:266]:
```suggestion
        if (isPaginationRequired(page,size)) {
            trustList = trustListMapper.trustListToTrustListDto(
                trustListService.getTrustList(mappedType, ifModifiedSince, page, size));
        } else {
            trustList = trustListMapper.trustListToTrustListDto(
                trustListService.getTrustList(mappedType, ifModifiedSince, null, null));
        }
```

File: src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java [hunk 11]
```diff
@@ -204,18 +298,41 @@
                 name = "country",
                 description = "2-Digit Country Code to filter for",
                 example = "EU",
-                required = true)
+                required = true),
+            @Parameter(
+                in = ParameterIn.HEADER,
+                name = "If-Modified-Since",
```
Comment (f11h) [inline src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java:304]:
```suggestion
                name = HttpHeaders.IF_MODIFIED_SINCE,
```

File: src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java [hunk 12]
```diff
@@ -231,14 +348,31 @@
     public ResponseEntity<List<TrustListDto>> downloadTrustListFilteredByCountryAndType(
         @Valid @PathVariable("type") CertificateTypeDto type,
         @Valid @Size(max = 2, min = 2) @PathVariable("country") String countryCode,
+        @RequestHeader(value = IF_MODIFIED_SINCE_HEADER, required = false) String ifModifiedSince,
```
Comment (f11h) [inline src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java:351]:
```suggestion
        @RequestHeader(value = HttpHeaders.IF_MODIFIED_SINCE, required = false) ZonedDateTime ifModifiedSince,
```

File: src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java [hunk 13]
```diff
@@ -231,14 +348,31 @@
...
+                return ResponseEntity.status(HttpStatus.BAD_REQUEST).build();
+            }
+        }
+        if (isPaginationRequired(page,size)) {
+            trustList = trustListMapper.trustListToTrustListDto(
+                trustListService.getTrustList(mappedType, countryCode, ifModifiedDateTime, page, size));
+        } else {
+            trustList = trustListMapper.trustListToTrustListDto(
+                trustListService.getTrustList(mappedType, countryCode, ifModifiedDateTime, null, null));
+        }
```
Comment (f11h) [inline src/main/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListController.java:375]:
```suggestion
        if (isPaginationRequired(page,size)) {
            trustList = trustListMapper.trustListToTrustListDto(
                trustListService.getTrustList(mappedType, countryCode, ifModifiedSince, page, size));
        } else {
            trustList = trustListMapper.trustListToTrustListDto(
                trustListService.getTrustList(mappedType, countryCode, ifModifiedSince, null, null));
        }
```

**Review by f11h — CHANGES_REQUESTED** [review]

**Review by ascheibal — APPROVED** [review]

**Review by ascheibal — APPROVED** [review]

**Review by ascheibal — APPROVED** [review]

File: src/test/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListIntegrationTest.java [hunk 14]
```diff
@@ -95,8 +100,17 @@
...
-
-    X509Certificate certUploadDe, certUploadEu, certCscaDe, certCscaEu, certAuthDe, certAuthEu, certDscDe, certDscEu;
+    private static final String IF_MODIFIED_SINCE_HEADER = "If-Modified-Since";
+    private static final ZoneId gmt = ZoneId.of("GMT");
+    private static final ZonedDateTime now = ZonedDateTime.now(gmt);
+    private static final ZonedDateTime nowMinusOneMinute = ZonedDateTime.now(gmt).minusMinutes(1);
+    private static final ZonedDateTime nowMinusOneHour = ZonedDateTime.now(gmt).minusHours(1);
+    private static final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("EEE, dd MMM yyyy HH:mm:ss zzz");
+    private static final String nowMinusOneMinuteStr = formatter.format(nowMinusOneMinute);
+    private static final String nowMinusOneHourStr = formatter.format(nowMinusOneHour);
```
Comment (f11h) [inline src/test/java/eu/europa/ec/dgc/gateway/restapi/controller/TrustListIntegrationTest.java:110]: This has to be ISO 8601 date format

**Review by f11h — APPROVED** [review]

**Pull request merged**